### PR TITLE
Fixing JCloudsApiExtensionPoint.getProvider

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-51460](https://issues.jenkins-ci.org/browse/JENKINS-51460)

In `/configure`:

screenshot

In `jenkins.model.ArtifactManagerConfiguration.xml`:

```xml
<?xml version='1.1' encoding='UTF-8'?>
<jenkins.model.ArtifactManagerConfiguration>
  <artifactManagerFactories>
    <io.jenkins.plugins.artifact__manager__s3.JCloudsArtifactManagerFactory plugin="artifact-manager-s3@…">
      <provider class="io.jenkins.plugins.artifact_manager_s3.S3BlobStore"/>
    </io.jenkins.plugins.artifact__manager__s3.JCloudsArtifactManagerFactory>
  </artifactManagerFactories>
</jenkins.model.ArtifactManagerConfiguration>
```

In `build.xml`:

```xml
  <artifactManager class="io.jenkins.plugins.artifact_manager_s3.JCloudsArtifactManager" plugin="artifact-manager-s3…">
    <provider class="io.jenkins.plugins.artifact_manager_s3.S3BlobStore"/>
  </artifactManager>
```

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/artifact_manager_s3/JCloudsArtifactManager.java [hunk 1]
```diff
@@ -66,29 +62,26 @@
...
  */
-class JCloudsArtifactManager extends ArtifactManager implements StashManager.StashAwareArtifactManager {
+final class JCloudsArtifactManager extends ArtifactManager implements StashManager.StashAwareArtifactManager {
 
     private static final Logger LOGGER = Logger.getLogger(JCloudsArtifactManager.class.getName());
 
-    private static String PROVIDER = System.getProperty("jclouds.provider", "aws-s3");
-
-    private static String BLOB_CONTAINER = System.getenv("S3_BUCKET");
-    private static String PREFIX = System.getenv("S3_DIR");
```
Comment (jglick) [inline src/main/java/io/jenkins/plugins/artifact_manager_s3/JCloudsArtifactManager.java]: Note that this would require trivial changes to #20.

Comment (oleg-nenashev) [reply src/main/java/io/jenkins/plugins/artifact_manager_s3/JCloudsArtifactManager.java]: Fine with me. I am waiting for fixes for `runATH()` there

**Comment by jglick** [comment]
Putting on hold for a moment while I think through the implications of having `JCloudsArtifactManager` store a `BlobStoreProvider` directly, rather than treating `BlobStoreProvider` as a factory. Does not matter for the current `S3BlobStore` but might if we chose to make the container and/or prefix fields be configurable.

**Comment by carlossg** [comment]
trying it I got 

```
java.lang.NullPointerException
	at io.jenkins.plugins.artifact_manager_s3.JCloudsArtifactManager.getContext(JCloudsArtifactManager.java:277)
	at io.jenkins.plugins.artifact_manager_s3.JCloudsArtifactManager.archive(JCloudsArtifactManager.java:105)
	at hudson.tasks.ArtifactArchiver.perform(ArtifactArchiver.java:235)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:80)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:67)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution$1$1.call(SynchronousNonBlockingStepExecution.java:49)
	at hudson.security.ACL.impersonate(ACL.java:290)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution$1.run(SynchronousNonBlockingStepExecution.java:46)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

```

**Comment by jglick** [comment]
@carlossg probably from some old build. Try with a clean working directory.

**Comment by kuisathaverat** [comment]
@jglick Would we have more cloud providers than S3? if so, the name would be weird artifact-manager-s3-plugin

**Comment by jglick** [comment]
> Would we have more cloud providers than S3?

Perhaps at some point, based on Azure etc. The SPI is still marked `@Beta` so we are not soliciting implementations from unknown parties; my immediate motivation is rather to clean up the code base and ease testing in #28.

> the name would be weird artifact-manager-s3-plugin

If and when we decide that the extension point is something we would like to officially advertise for other (non-test) implementations, we would want to split out an `artifact-manager-jclouds` plugin. I had considered actually renaming packages to that effect now, and had not bothered, but perhaps it is clearest and safest to do this now (otherwise you need to make an XStream `alias` call later for settings compatibility).

File: src/main/java/io/jenkins/plugins/artifact_manager_s3/S3BlobStore.java [hunk 2]
```diff
@@ -60,37 +62,45 @@
...
-public class S3BlobStore extends JCloudsApiExtensionPoint {
+public class S3BlobStore extends BlobStoreProvider {
 
     private static final Logger LOGGER = Logger.getLogger(S3BlobStore.class.getName());
 
     private static final long serialVersionUID = -8864075675579867370L;
 
+    // For now, these are taken from the environment, rather than being configured.
+    @SuppressWarnings("FieldMayBeFinal")
+    private static String BLOB_CONTAINER = System.getenv("S3_BUCKET");
```
Comment (carlossg) [inline src/main/java/io/jenkins/plugins/artifact_manager_s3/S3BlobStore.java]: why make them S3 specific when affect any future providers too?

Comment (jglick) [reply src/main/java/io/jenkins/plugins/artifact_manager_s3/S3BlobStore.java]: Just to be on the conservative side—we do not know yet if configuration of container and/or prefix will be desirable or necessary for other cloud providers. This API allows the provider to decide whether to hard-code values; compute them from some more general settings; get them from environment variables (system properties, …); or let the user select them in `/configure`, presumably with form validation and completion and so on, possibly alongside credentials.

**Pull request merged**